doc/custom-metrics-adapter.md

```diff
@@ -0,0 +1,77 @@
+# Custom metrics Adapter
+
+This adapter contains an implementation of [external metrics](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/external-metrics-api.md)
+ API. It is therefore suitable for use with the autoscaling/v2 Horizontal Pod Autoscaler in Kubernetes 1.6+.
+ 
+
+## Use kustomize to customise your deployment
+
+1. Clone the source code:
+
+```sh
+git clone git@github.com:apache/skywalking-swck.git
+```
+
+2. Edit file `config/adapter/adapter/kustomization.yaml` file to change your preferences. If you prefer to your private 
+ docker image, a quick path to override `ADAPTER_IMG` environment variable : `export ADAPTER_IMG=<private registry>/adapter:<tag>`
+
+3. Use `make` to generate the final manifests and deploy:
+
+```sh
+make adapter-deploy
+```
+
+## Configuration
+
+The adapter takes the standard Kubernetes generic API server arguments (including those for authentication and authorization). 
+By default, it will attempt to using Kubernetes in-cluster config to connect to the cluster.
+
+It takes the following addition arguments specific to configuring how the adapter talks to SkyWalking OAP cluster:
+
+ * `--oap-addr` The address of OAP cluster.
+ * `--metric-filter-regex` A regular expression to filter metrics retrieved from OAP cluster.
+ * `--refresh-interval` This is the interval at which to update the cache of available metrics from OAP cluster. 
+ 
+## HPA Configuration
+
+External metrics allow you to autoscale your cluster based on any metric available in OAP cluster. 
+ Just provide a metric block with a `name` and `selector`, and use the `External` metric type.
+ 
+```yaml
+- type: External
+  external:
+    metric:
+      name: <metric_name>
+      metricSelector:
+        matchLabels:
+          <label_key>: <label_value>
+          ...
+    target:
+      ....
+```
+
+ * metric_name: The name of metric generated by OAL or other subsystem.
+ * label: available `label_key`s are `service` and `label`. 
+  
+ The options of label keys are:
+ * `service` varies in different scopes. If the scope is `Service`, `ServiceInstance` and `Endpoint`, the `service` represents
+    `serviceName`, `serviceInstanceName` and `endpointName` respectively.
```

Comment:
How to map to ServiceInstance and EndpointName by a string name? Including how about the relation metrics, such as ServiceRelation.

Reply:
>How to map to ServiceInstance and EndpointName by a string name? 

How about adding a link to swctl relevant sections in its README? 
 * linear: https://github.com/apache/skywalking-cli#metrics-linear
 * multi-lines : https://github.com/apache/skywalking-cli#metrics-linear, but I use a string `label` instead of a natural number `num` to support meter system.

> Including how about the relation metrics, such as ServiceRelation.

have no plan to implement it. HPA cares about the node metrics instead of the relation ones. Do you have any scenarios for it? feel free to let me know.

Reply:
> $ ./bin/swctl --display=graph metrics linear --name=service_instance_resp_time --service "load balancer1.system"

This is how to query `instance` level metrics, which I can't find a way to map your documentation.

> have no plan to implement it. HPA cares about the node metrics instead of the relation ones. Do you have any scenarios for it? feel free to let me know.

Make sense to me.

Reply:
The adapter is able to opt to the below pieces to query OAP just as the swctl command.
```
- type: External
  external:
    metric:
      name: service_instance_resp_time
      metricSelector:
        matchLabels:
          service: load balancer1.system
          ...
```
the adapter doesn't need a `display` argument, and it could analyze metric type(linear/multiple-lines) from the `listMetric` query of GraphQL.

Reply:
https://github.com/apache/skywalking/issues/5890 I think this caused your misunderstanding. Let's hold this PR for a while, we should wait for CLI 0.5.0. Agree?

Reply:
Noop. swck doesn't depend on a release. once this gets merged, swck could sync with it.

Reply:
Then, we need a deeper discussion about what do you plan to distribute. In the current release on the download page, http://skywalking.apache.org/downloads/, there is only a source release, so we always expect users to compile codes by themselves. How would this adaptor be?

Reply:
> there is only a source release

We release a binary package actually, https://dist.apache.org/repos/dist/release/skywalking/swck/0.1.0/. But it's used to build the release image. The binary can't be run outside kubernetes, that's the reason I didn't add a link to our website. In the next release/0.2.0, I will add adapter binary into it.

The image release strategy is since binary release package contains `operator` and `adapter`, I will add them to a single image called `apache/skywalking-swck`. The default entry point of the image is `operator`, if uses want to run adapter, they should add `command: /bin/adapter` to the pod container node. 

You might notice that there are two different images for the operator and adapter respectively. They are only for dev purposes. The docker file of the release version is going to be submitted soon.

Reply:
> We release a binary package actually, https://dist.apache.org/repos/dist/release/skywalking/swck/0.1.0/. But it's used to build the release image.

This is strange, we have binary on the apache mirror system, but no one is able to download it, and no one knows it exists. I think we need to fix this.

Image is another topic, if you want to release that, we need the docker file in the release binary tar ball.